collect_app/src/main/java/org/odk/collect/android/formlists/blankformlist/BlankFormListActivity.kt

```diff
@@ -0,0 +1,144 @@
+import javax.inject.Inject
+
+class BlankFormListActivity : LocalizedActivity(), OnFormItemClickListener {
+
+    @Inject
+    lateinit var viewModelFactory: BlankFormListViewModel.Factory
+
+    @Inject
+    lateinit var networkStateProvider: NetworkStateProvider
+
+    @Inject
+    lateinit var permissionsProvider: PermissionsProvider
+
+    private val viewModel: BlankFormListViewModel by viewModels { viewModelFactory }
+
+    private val adapter: FormListAdapter = FormListAdapter().apply {
+        listener = this@BlankFormListActivity
+    }
+
+    private lateinit var menuDelegate: BlankFormListMenuDelegate
+
+    override fun onCreate(savedInstanceState: Bundle?) {
+        super.onCreate(savedInstanceState)
+        DaggerUtils.getComponent(this).inject(this)
+        setContentView(R.layout.activity_blank_form_list)
+        title = getString(R.string.enter_data)
+        setSupportActionBar(findViewById(R.id.toolbar))
+
+        menuDelegate = BlankFormListMenuDelegate(this, viewModel, networkStateProvider)
+
+        findViewById<RecyclerView>(R.id.formList).adapter = adapter
+
+        initObservers()
+    }
+
+    override fun onCreateOptionsMenu(menu: Menu): Boolean {
+        menuDelegate.onCreateOptionsMenu(menuInflater, menu)
+        return true
+    }
+
+    override fun onPrepareOptionsMenu(menu: Menu): Boolean {
+        menuDelegate.onPrepareOptionsMenu(menu)
+        return true
+    }
+
+    override fun onOptionsItemSelected(item: MenuItem): Boolean {
+        return menuDelegate.onOptionsItemSelected(item)
+    }
+
+    override fun onFormClick(formUri: Uri) {
+        if (Intent.ACTION_PICK == intent.action) {
+            // caller is waiting on a picked form
+            setResult(RESULT_OK, Intent().setData(formUri))
+        } else {
+            // caller wants to view/edit a form, so launch formentryactivity
+            Intent(this, FormEntryActivity::class.java).apply {
+                action = Intent.ACTION_EDIT
+                data = formUri
+                putExtra(
+                    ApplicationConstants.BundleKeys.FORM_MODE,
+                    ApplicationConstants.FormModes.EDIT_SAVED
+                )
+
+                startActivity(this)
+            }
+        }
+        finish()
+    }
+
+    override fun onMapButtonClick(id: Long) {
+        permissionsProvider.requestEnabledLocationPermissions(
+            this,
+            object : PermissionListener {
+                override fun granted() {
+                    startActivity(
+                        Intent(this@BlankFormListActivity, FormMapActivity::class.java).also {
+                            it.putExtra(FormMapActivity.EXTRA_FORM_ID, id)
+                        }
+                    )
+                }
+
+                override fun denied() = Unit
+            }
+        )
+    }
+
+    private fun initObservers() {
+        viewModel.backgroundTasksObserver.observe(this) {
```

Comment:
Instead of just zipping the three `LiveData` objects together you could zip3 them then do a `Transformations.map` to get a `LiveData<Boolean>`:

```kotlin
val isLoading: LiveData<Boolean> = Transformations.map(LiveDataUtils.zip3(
    isFormLoadingRunning,
    isSyncingWithStorageRunning,
    isSyncingWithServerRunning,
)) { (one, two, three) -> one || two || three  }
```

Alternatively, I think you could use a count and a `LiveData<Boolean` (like a counting lock):

```kotlin

private var loadingCount = 0
val isLoading: LiveData<Boolean> = MutableLiveData<Boolean>(false)

private fun loadFromDatabase() {
    startLoading()
    scheduler.immediate(
        foreground = { .. },
        background = {
             finishLoading()
        }
    )
}

private fun startLoading() {
    loadingCount = loadingCount + 1
    isLoading.value = loadingCount > 0
}

private fun finishLoading() {
    loadingCount = loadingCount - 1
    isLoading.value = loadingCount > 0
}

Reply:
Done.